### PR TITLE
New feature ... Dynamic Tabs :sparkles:

### PR DESCRIPTION
**Pull request by jbbjarnason**
Tabs can by created and added separately to the web.

I refactored a bit the json part of this project.

Great work initiating this awesome dashboard.

### CODE REVIEW

**Comment by ayushsharma82** [comment]
Nice addition! I'll have a look at it when I get some time off my busy schedule.

In the meanwhile, here are some tests to be performed before the merger:

- [ ] Doesn't has any performance impact
- [ ] All cards show up on webpage
- [ ] All charts show up on webpage
- [ ] Dynamic addition/removal of cards & charts
- [ ] Card updates work without page refresh
- [ ] Chart updates work without page refresh

All kudos to you for the wonderful addition, but a part of me do feel that this could be have been done in a simpler way without so much C++ overhead? We could have a `_tab` `String` variable in each card & chart with the name of tab you want to put it in. Then, later on, all the sorting would have been done in the webpage where the js part of our webpage would sort all the devices depending on the name of the tag and at the same time add the tabs in the navbar.

Let me know your thoughts on this.

File: src/Card.cpp [hunk 1]
```diff
@@ -1,22 +1,46 @@
...
-  #endif
+  _tab = nullptr;
   _type = type;
   _name = name;
   _symbol = symbol;
   _value_min = min;
   _value_max = max;
   _dashboard->add(this);
 }
+Card::Card(Tab *tab, const int type, const char* name, const char* symbol, const int min, const int max):
```
Comment (jbbjarnason) [inline src/Card.cpp:32]: Maybe the pointer could be of type etl::variant<ESPDash*, Tab*> and reuse the code.

**Comment by ayushsharma82** [comment]
Awesome, I get your point now. I plan on maintaining backwards compatibility as this is not a major release, so for that reason namespacing various parts of ESPDash is not suitable. 

Secondly, Can you describe `Widget` in the code?

Here are the conditions for minor releases:
- [x] - No extra dependencies
- [ ] - Major functionality should remain the same ( including constructors )
- [ ] - Any removed functions should be marked as deprecated and emptied out

Some suggestions:
- Existing class constructors should not be modified, that will lead to users re-learning the syntax of this library
- Create a tab storage class just like card/chart where its constructor will take a pointer of ESPDash class.
- Add `setTab` function in card/chart class to assign tab ( card/chart should store a pointer to tab storage class ).
- All cards/charts for which user haven't supplied "tab" via `setTab` function, those should be placed in "Overview" page of our webpage. ( You can check that via `nullptr` ) ( ie, all unsigned cards should be put in "Overview" page ).

I'll review your additions in the upcoming days. ( might take some time, I'm stuck in between some large projects ). Till then, You can add to the pull request which you think is necessary.

File: src/ESPDash.cpp [hunk 2]
```diff
@@ -104,251 +29,175 @@ void ESPDash::setAuthentication(const char *user, const char *pass) {
...
+              serializeJson(generateLayout(&id), response);
+            else
+              serializeJson(generateLayout(), response);
+          }
+          else if (command == "ping")
+            response = "{\"command\":\"pong\"}";
+          else if (command == "getStats")
+            response = generateStatsJSON();
+          else if (command == "buttonClicked") {
+            home_screen->resolveCardCallback(
```
Comment (jbbjarnason) [inline src/ESPDash.cpp:116]: this is was forgotten, needs to be fixed

Comment (jbbjarnason) [reply src/ESPDash.cpp]: is fixed in latest commit

File: src/Card.cpp (hunk 1 above)
Comment (jbbjarnason) [reply src/Card.cpp:32]: Changes the constructor, hence will not do

**Comment by jbbjarnason** [comment]
> Awesome, I get your point now. I plan on maintaining backwards compatibility as this is not a major release, so for that reason namespacing various parts of ESPDash is not suitable.
> 
> Secondly, Can you describe `Widget` in the code?
> 
> Here are the conditions for minor releases:
> 
> * [x]  - No extra dependencies
> * [ ]  - Major functionality should remain the same ( including constructors )
> * [ ]  - Any removed functions should be marked as deprecated and emptied out
> 
> Some suggestions:
> 
> * Existing class constructors should not be modified, that will lead to users re-learning the syntax of this library
> * Create a tab storage class just like card/chart where its constructor will take a pointer of ESPDash class.
> * Add `setTab` function in card/chart class to assign tab ( card/chart should store a pointer to tab storage class ).
> * All cards/charts for which user haven't supplied "tab" via `setTab` function, those should be placed in "Overview" page of our webpage. ( You can check that via `nullptr` ) ( ie, all unassigned cards should be put in "Overview" page ).
> 
> I'll review your additions in the upcoming days. ( might take some time, I'm stuck in between some large projects ). Till then, You can add to the pull request which you think is necessary.


Widget is just a abstraction of Card or Chart, I feel like they could be made more general. At least for now they share some code and I really dislike duplicated code. Therefore, I pulled some variables into a Widget class.

Major functionality is the same, however, the statistics button only appears now if it is enabled. Otherwise it is hidden.

I haven't removed any public function, but some of the private ones where removed/changed.

Answering the suggestions:
- The class constructors should be the same.
- Tab class has a pointer to ESPDash, like the card&chart.
- Not really a fan of a setTab function since it the API is more like the same as before, you have now a possibility to make a tab and it is created like a card&chart and can be used like a secondary dashboard, when creating a card&chart inside the tab.
- The last suggestion would require the ESPDash class to have ownership(pointer) to all the card&chart objects, I really don't see why you would want to have it like so? Currently all card&chart which are created with reference to the ESPDash class appear on the Overview page, if you create a card/chart with reference to a tab it appears in that tab.

I would recommend to review the example code first, as it explains how the public API for this new feature is. 
https://github.com/jbbjarnason/ESP-DASH/blob/master/examples/Tabs/Tabs.ino#L53

I hope I am not too aggressive, sorry if you feel like so.

**Comment by jbbjarnason** [comment]
I will let you know when I have finished. I need at least to finish those things:

- [ ]  Headers
- [ ]  Integration with esphome

and any upcoming issues that arise

**Comment by jbbjarnason** [comment]
I will not be making any more changes in this PR, only those you would request me to change. So it is ready for review.

If you are interested in esphome, I have finished the [code generator](https://github.com/jbbjarnason/esphome-dashboard)

**Comment by jbbjarnason** [comment]
closed due to updating my own fork master, since it hasn't been merged

**Comment by ayushsharma82** [comment]
I've tested the functionality and it works perfectly but I need to merge it with the pro version as well otherwise there will be a huge discrepancy between the 2 versions. 

I'll take note of your last commit in this PR and credit you for your work when it's merged. I'll just need some time to do that between my work schedule. Thanks a lot for the effort!

**Comment by jbbjarnason** [comment]
> I've tested the functionality and it works perfectly but I need to merge it with the pro version as well otherwise there will be a huge discrepancy between the 2 versions.
> 
> I'll take note of your last commit in this PR and credit you for your work when it's merged. I'll just need some time to do that between my work schedule. Thanks a lot for the effort!

Great! I will be doing some changes in my fork, for example asynchronous updates for widgets and row/column impl for each tab. But it will break the public interface, so I am not sure if I will do a second PR based on that. Good luck, great project!

**Pull request closed without merging**